### PR TITLE
JetBrains: migrate `ClientInfo`, `ConfigUtil` & `ExtensionConfiguration` to Kotlin

### PR DESCRIPTION
**Pull request by Strum355**
Migration of some length pojos to nicer Kotlin equivalents. Includes some drive-by stylistic improvements, as well as aligning some types that cross the agent boundary with https://sourcegraph.com/github.com/sourcegraph/cody/-/blob/agent/src/protocol.ts

## Test plan

Compilation still works, Cody Agent still accepts all types

### CODE REVIEW

File: client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      config.autocompleteAdvancedProvider = it.vscodeSettingString()
+    }
+
+    return config
+  }
+
+  @JvmStatic
+  fun getConfigAsJson(project: Project): JsonObject {
+    val (instanceUrl, accessToken, customRequestHeaders) = ServerAuthLoader.loadServerAuth(project)
+    val (_, _, _, _, _, _, anonymousUserId) = CodyApplicationSettings.getInstance()
```
Comment (szymonprz) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:59]: maybe it's better not to deconstruct this? we shouldn't have compilation problems when we add a new field to the settings

File: client/jetbrains/src/main/java/com/sourcegraph/cody/agent/CodyAgent.kt [hunk 2]
```diff
@@ -59,18 +59,15 @@ class CodyAgent(private val project: Project) : Disposable {
...
           val info =
               server
                   .initialize(
-                      ClientInfo()
-                          .setName("JetBrains")
-                          .setVersion(ConfigUtil.getPluginVersion())
-                          .setWorkspaceRootPath(ConfigUtil.getWorkspaceRoot(project))
-                          .setExtensionConfiguration(ConfigUtil.getAgentConfiguration(project)))
+                      ClientInfo(
+                          name = "JetBrains",
```
Comment (szymonprz) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/agent/CodyAgent.kt:67]: i love it!

File: client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import com.sourcegraph.cody.config.CodyAuthenticationManager
+import com.sourcegraph.cody.config.ServerAuthLoader
+import com.sourcegraph.cody.config.SourcegraphServerPath
+import com.sourcegraph.cody.config.SourcegraphServerPath.Companion.from
+import java.util.*
+import java.util.stream.Collectors
+import org.jetbrains.annotations.Contract
+
+object ConfigUtil {
+  const val DOTCOM_URL = "https://sourcegraph.com/"
```
Comment (Strum355) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:24]: [this test](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/jetbrains/src/test/kotlin/com/sourcegraph/cody/config/SourcegraphServerPathTest.kt?L10) is sensitive to the `/` being at the end :upside_down_face:

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/agent/ExtensionConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.sourcegraph.cody.agent
+
+data class ExtensionConfiguration
+internal constructor(
```
Comment (szymonprz) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/cody/agent/ExtensionConfiguration.kt:4]: do we need to mark this constructor as internal?

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/BaseLoginDialog.kt [hunk 5]
```diff
@@ -23,18 +23,14 @@ abstract class BaseLoginDialog(
 
   protected val loginPanel = CodyLoginPanel(executorFactory, isAccountUnique)
 
-  private var _login = ""
-  private var _displayName: String? = null
-  private var _token = ""
+  var login: String = ""
+    private set
```
Comment (szymonprz) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/BaseLoginDialog.kt:27]: nice idea :)

**Review by szymonprz — APPROVED** [review]

File: client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt (hunk 1 above)
Comment (Strum355) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:59]: Yea was debating whether I should change this or not, probably best to do so. Result of intellijs java to kotlin conversion :smile:

File: client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt (hunk 3 above)
Comment (szymonprz) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:24]: yeah, I think that somewhere in our clients to the backend we require to have "/" at the end, but maybe it's not true anymore

File: client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt (hunk 1 above)
Comment (olafurpg) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:59]: > we shouldn't have compilation problems when we add a new field to the settings

Big +1

Comment (olafurpg) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/config/ConfigUtil.kt:59]: Another big problem: renaming the original symbol doesn't always rename the variable from the pattern. It's better to access the fields directly via `a.b` syntax.

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/CodyAccountManager.kt [hunk 6]
```diff
@@ -8,7 +8,7 @@ import com.sourcegraph.config.ConfigUtil
 @Service
 class CodyAccountManager :
     AccountManagerBase<CodyAccount, String>(ConfigUtil.SERVICE_DISPLAY_NAME) {
-  override fun accountsRepository() = service<CodyPersisentAccounts>()
+  override fun accountsRepository() = service<CodyPersistentAccounts>()
```
[the remark this answers is not in the document]
Comment (vdavid) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/CodyAccountManager.kt:11]: Thank you! This hurt my eyes before! <3

**Pull request merged**